File: 29Items-Projects/7Algorithms/6-Backtracking/python/palindrome_partitioning.py
"""Palindrome Partitioning — split string into substrings where each is a palindrome."""


class PalindromePartitioning:
    def __init__(self, s: str):
        if not s:
            raise ValueError("String must not be empty")
        self.s = s
        self.solutions: list[list[str]] = []

    def solve_and_print(self) -> None:
        self._backtrack([], 0)
        print(f"Found {len(self.solutions)} partition(s) for \"{self.s}\"\n")

        limit = min(len(self.solutions), 6)
        for i, sol in enumerate(self.solutions[:limit], 1):
            print(f"#{i}: {sol}")
        if len(self.solutions) > limit:
            print(f"... and {len(self.solutions) - limit} more.")

    def _backtrack(self, current: list[str], start: int) -> None:
        if start == len(self.s):
            self.solutions.append(current[:])
            return
        for end in range(start + 1, len(self.s) + 1):
            sub = self.s[start:end]
            if sub == sub[::-1]:
                current.append(sub)
                self._backtrack(current, end)
                current.pop()

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/palindromepartitioning/PalindromePartitioning.java
package palindromepartitioning;

import java.util.*;

public class PalindromePartitioning {

    private final String s;
    private final List<List<String>> solutions = new ArrayList<>();

    public PalindromePartitioning(String s) {
        if (s == null || s.isEmpty())
            throw new IllegalArgumentException("String must not be empty");
        this.s = s;
    }

    public void solveAndPrint() {
        backtrack(new ArrayList<>(), 0);
        System.out.printf("Found %d partition(s) for \"%s\"%n%n", solutions.size(), s);

        int limit = Math.min(solutions.size(), 6);
        for (int i = 0; i < limit; i++) {
            System.out.printf("#%d: %s%n", i + 1, solutions.get(i));
        }
        if (solutions.size() > limit) {
            System.out.printf("... and %d more.%n", solutions.size() - limit);
        }
    }

    private void backtrack(List<String> current, int start) {
        if (start == s.length()) {
            solutions.add(new ArrayList<>(current));
            return;
        }
        for (int end = start + 1; end <= s.length(); end++) {
            String sub = s.substring(start, end);
            if (isPalindrome(sub)) {
                current.add(sub);
                backtrack(current, end);
                current.remove(current.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String str) {
        int l = 0, r = str.length() - 1;
        while (l < r) {
            if (str.charAt(l++) != str.charAt(r--)) return false;
        }
        return true;
    }
}
